File: examples/java/local/app/src/androidTest/java/com/mongodb/realm/examples/java/MigrateFromJavaToKotlinSDKTest.java
package com.mongodb.realm.examples.java;

import android.util.Log;

import com.mongodb.realm.examples.Expectation;
import com.mongodb.realm.examples.RealmTest;
import com.mongodb.realm.examples.model.java.SampleJava;

import org.junit.Test;

import java.util.Arrays;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import io.realm.Realm;
import io.realm.RealmConfiguration;
import io.realm.RealmResults;
import io.realm.RealmSchema;
import io.realm.Sort;
import io.realm.exceptions.RealmFileException;
// :replace-start: {
//    "terms": {
//       "KidJava": "Kid",
//       "SampleJava": "Sample",
//       "ChildJava": "Child",
//       "CollegeStudentJava": "CollegeStudent"
//    }
// }
public class MigrateFromJavaToKotlinSDKTest extends RealmTest {
    @Test
    public void testOpenARealm() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            // :code-block-start: open-a-realm
            RealmConfiguration config =
                    new RealmConfiguration.Builder()
                    .build();

            Realm realm;
            realm = Realm.getInstance(config);
            Log.v("EXAMPLE",
            "Successfully opened a realm: "
                    + realm.getPath());
            // :code-block-end:
            realm = Realm.getInstance(config);
            realm.close();
            expectation.fulfill();
        });
        expectation.await();
    }

    @Test
    public void testAsyncWrite() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            RealmConfiguration config =
                    new RealmConfiguration.Builder()
                            .name(getRandom())
                            .allowQueriesOnUiThread(true)
                            .allowWritesOnUiThread(true)
                            .build();

            Realm realm;
            try {
                realm = Realm.getInstance(config);
                Log.v("EXAMPLE",
                        "Successfully opened a realm: "
                                + realm.getPath());
                // :code-block-start: write-async
                realm.executeTransactionAsync(
                        transactionRealm -> {
                    SampleJava sample = new SampleJava();
                    sample.stringField = "Sven";
                    transactionRealm.copyToRealm(sample);
                });
                // :code-block-end:
            } catch (RealmFileException ex) {
                Log.v("EXAMPLE",
                        "Error opening the realm.");
                Log.v("EXAMPLE",
                        ex.toString());
            }
            realm = Realm.getInstance(config);
            realm.close();
            expectation.fulfill();
        });
        expectation.await();
    }

    @Test
    public void testSyncWrite() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            RealmConfiguration config =
                    new RealmConfiguration.Builder()
                            .name(getRandom())
                            .allowQueriesOnUiThread(true)
                            .allowWritesOnUiThread(true)
                            .build();

            Realm realm;
            try {
                realm = Realm.getInstance(config);
                // :code-block-start: write-sync
                realm.executeTransaction(
                        transactionRealm -> {
                    SampleJava sample = new SampleJava();
                    sample.stringField = "Sven";
                    transactionRealm.copyToRealm(sample);
                });
                // :code-block-end:
                Log.v("EXAMPLE",
                        "Successfully opened a realm: "
                                + realm.getPath());
            } catch (RealmFileException ex) {
                Log.v("EXAMPLE",
                        "Error opening the realm.");
                Log.v("EXAMPLE",
                        ex.toString());
            }
            realm = Realm.getInstance(config);
            realm.close();
            expectation.fulfill();
        });
        expectation.await();
    }

    @Test
    public void testQuery() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            RealmConfiguration config =
                    new RealmConfiguration.Builder()
                            .name(getRandom())
                            .allowQueriesOnUiThread(true)
                            .allowWritesOnUiThread(true)
                            .build();

            Realm realm;
            try {
                realm = Realm.getInstance(config);
                // :code-block-start: query-filters
                RealmResults<SampleJava> samples =
                        realm
                            .where(SampleJava.class)
                            .findAll();

                RealmResults<SampleJava> samplesThatBeginWithN =
                        realm
                            .where(SampleJava.class)
                            .beginsWith("stringField",
                                    "N")
                            .findAll();
                // :code-block-end:
                Log.v("EXAMPLE",
                        "Successfully opened a realm: "
                                + realm.getPath());
            } catch (RealmFileException ex) {
                Log.v("EXAMPLE",
                        "Error opening the realm.");
                Log.v("EXAMPLE",
                        ex.toString());
            }
            realm = Realm.getInstance(config);
            realm.close();
            expectation.fulfill();
        });
        expectation.await();
    }

    @Test
    public void testQuerySortLimitDistinct() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            RealmConfiguration config =
                    new RealmConfiguration.Builder()
                            .name(getRandom())
                            .allowQueriesOnUiThread(true)
                            .allowWritesOnUiThread(true)
                            .build();

            Realm realm;
            try {
                realm = Realm.getInstance(config);
                // :code-block-start: query-sort-limit-distinct
                RealmResults<SampleJava> aggregates =
                        realm.where(SampleJava.class)
                        .distinct("stringField")
                        .sort("stringField",
                                Sort.ASCENDING)
                        .limit(2)
                        .findAll();
                // :code-block-end:
                Log.v("EXAMPLE",
                        "Successfully opened a realm: "
                                + realm.getPath());
            } catch (RealmFileException ex) {
                Log.v("EXAMPLE",
                        "Error opening the realm.");
                Log.v("EXAMPLE",
                        ex.toString());
            }
            realm = Realm.getInstance(config);
            realm.close();
            expectation.fulfill();
        });
        expectation.await();
    }

    @Test
    public void testDeletes() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            RealmConfiguration config =
                    new RealmConfiguration.Builder()
                            .name(getRandom())
                            .allowQueriesOnUiThread(true)
                            .allowWritesOnUiThread(true)
                            .build();

            Realm realm;
            try {
                realm = Realm.getInstance(config);
                realm.executeTransaction(transactionRealm -> {
                    SampleJava sample = new SampleJava();
                    sample.stringField = "Sven";
                    transactionRealm.copyToRealm(sample);
                    sample.stringField = "not sven";
                    transactionRealm.copyToRealm(sample);
                });
                // :code-block-start: deletes
                SampleJava sample =
                        realm.where(SampleJava.class)
                                .findFirst();

                // delete one object synchronously
                realm.executeTransaction(
                        transactionRealm ->
                        sample.deleteFromRealm());

                // delete a query result asynchronously
                realm.executeTransactionAsync(
                        backgroundRealm ->
                        backgroundRealm
                                .where(SampleJava.class)
                                .findFirst()
                                .deleteFromRealm());
                // :code-block-end:
                Log.v("EXAMPLE",
                        "Successfully opened a realm: "
                                + realm.getPath());
            } catch (RealmFileException ex) {
                Log.v("EXAMPLE",
                        "Error opening the realm.");
                Log.v("EXAMPLE",
                        ex.toString());
            }
            realm = Realm.getInstance(config);
            realm.close();
            expectation.fulfill();
        });
        expectation.await();
    }

    @Test
    public void testNotifications() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            RealmConfiguration config =
                    new RealmConfiguration.Builder()
                            .name(getRandom())
                            .allowQueriesOnUiThread(true)
                            .allowWritesOnUiThread(true)
                            .build();

            Realm realm;
            try {
                realm = Realm.getInstance(config);
                realm.executeTransaction(new Realm.Transaction() {
                    @Override
                    public void execute(Realm realm) {
                        SampleJava sample = new SampleJava();
                        sample.stringField = "Sven";
                        realm.copyToRealm(sample);
                        sample.stringField = "not sven";
                        realm.copyToRealm(sample);
                    }
                });
                // :code-block-start: notifications
                realm.where(SampleJava.class).findAllAsync()
                    .addChangeListener(
                            (samples, changeSet) -> {
                        // log change description
                        Log.v("EXAMPLE",
                            "Results changed. " +
                            "change ranges: " +
                            Arrays.toString(
                                changeSet
                                    .getChangeRanges()) +
                            ", insertion ranges: " +
                            Arrays.toString(
                                changeSet
                                    .getInsertionRanges()) +
                            ", deletion ranges: " +
                            Arrays.toString(
                                changeSet
                                    .getDeletionRanges()));
                    });
                // :code-block-end:
                Log.v("EXAMPLE",
                        "Successfully opened a realm: "
                                + realm.getPath());
            } catch (RealmFileException ex) {
                Log.v("EXAMPLE",
                        "Error opening the realm.");
                Log.v("EXAMPLE",
                        ex.toString());
            }
            realm = Realm.getInstance(config);
            realm.close();
            expectation.fulfill();
        });
        expectation.await();
    }

    @Test
    public void testThreading() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            RealmConfiguration config =
                    new RealmConfiguration.Builder()
                            .name(getRandom())
                            .allowQueriesOnUiThread(true)
                            .allowWritesOnUiThread(true)
                            .build();

            Realm realm;
            try {
                // :code-block-start: threading
                realm = Realm.getInstance(config);
                // :hide-start:
                realm.executeTransaction(new Realm.Transaction() {
                    @Override
                    public void execute(Realm realm) {
                        SampleJava sample = new SampleJava();
                        sample.stringField = "Sven";
                        realm.copyToRealm(sample);
                        sample.stringField = "not sven";
                        realm.copyToRealm(sample);
                    }
                });
                // :hide-end:
                SampleJava sample = realm
                        .where(SampleJava.class).findFirst();
                // save sample field in a variable
                // for access on another thread
                String sampleStringField =
                        sample.stringField;
                ExecutorService executorService =
                        Executors.newFixedThreadPool(4);
                executorService.execute(() -> {
                    // cannot pass a realm
                    // into another thread,
                    // so get a new instance
                    // for separate thread
                    Realm threadRealm =
                            Realm.getInstance(config);
                    // cannot access original
                    // sample on another
                    // thread, use
                    // sampleStringField instead
                    SampleJava threadSample =
                            threadRealm
                                    .where(SampleJava.class)
                            .equalTo("stringField",
                                    sampleStringField)
                                    .findFirst();
                    Log.v("EXAMPLE",
                            "Separate thread sample: "
                            + threadSample);
                });
                // :code-block-end:
                Log.v("EXAMPLE",
                        "Successfully opened a realm: "
                                + realm.getPath());
            } catch (RealmFileException ex) {
                Log.v("EXAMPLE",
                        "Error opening the realm.");
                Log.v("EXAMPLE",
                        ex.toString());
            }
            realm = Realm.getInstance(config);
            realm.close();
            expectation.fulfill();
        });
        expectation.await();
    }

    @Test
    public void testMigration() {
        Expectation expectation = new Expectation();
        activity.runOnUiThread(() -> {
            // :code-block-start: migrations
            RealmConfiguration config =
                new RealmConfiguration.Builder()
                        .name(getRandom()) // :hide:
                        .migration((realm,
                                    oldVersion,
                                    newVersion) -> {
                    RealmSchema schema =
                            realm.getSchema();

                    if (oldVersion == 0L) {
                        // perform schema migration
                        schema.get("Sample")
                            .addField("new_field",
                                    String.class);
                    }

                    // migrate data
                    schema.get("Sample")
                        .transform(obj ->
                            obj.set("longField",
                                    42L));
                }).build();

            Realm realm;
            realm = Realm.getInstance(config);
            Log.v("EXAMPLE",
            "Successfully opened a realm: "
                    + realm.getPath());
            // :code-block-end:
            expectation.fulfill();
        });
        expectation.await();
    }
}
// :replace-end:

File: examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/MigrateFromJavaToKotlinSDKTest.kt
package com.mongodb.realm.realmkmmapp

import io.realm.Realm
import io.realm.RealmConfiguration
import io.realm.RealmInstant
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.RealmResults
import io.realm.annotations.Index
import io.realm.annotations.PrimaryKey
import io.realm.dynamic.DynamicMutableRealm
import io.realm.dynamic.DynamicMutableRealmObject
import io.realm.dynamic.DynamicRealm
import io.realm.dynamic.DynamicRealmObject
import io.realm.internal.platform.runBlocking
import io.realm.migration.AutomaticSchemaMigration
import io.realm.notifications.InitialResults
import io.realm.notifications.ResultsChange
import io.realm.notifications.UpdatedResults
import io.realm.query
import io.realm.query.Sort
import io.realm.realmListOf

import kotlin.test.Test
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

// :code-block-start: realm-object-model
class Sample : RealmObject {
    @PrimaryKey
    var stringField: String = "Realm"
    var byteField: Byte = 0xA
    var charField: Char = 'a'
    var shortField: Short = 17
    var intField: Int = 42
    @Index
    var longField: Long = 256L
    var booleanField: Boolean = true
    var floatField: Float = 3.14f
    var doubleField: Double = 1.19840122
    var timestampField: RealmInstant =
        RealmInstant.fromEpochSeconds(
            100,
            1000)
}
// :code-block-end:

// :code-block-start: one-to-one-relationship
class Child : RealmObject {
    var frog: Frog? = null
}
// :code-block-end:

// :code-block-start: one-to-many-relationship
class Kid : RealmObject {
    var frogs: RealmList<Frog> =
        realmListOf()
}
// :code-block-end:

// :code-block-start: schema-types
class Student : RealmObject {
    var notes: RealmList<String> =
        realmListOf()
    var nullableNotes: RealmList<String?> =
        realmListOf()
}
// :code-block-end:

class MigrateFromJavaToKotlinSDKTest: RealmTest() {
    @Test
    fun openARealmTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            // :code-block-start: open-a-realm
            val config = RealmConfiguration
                .with(schema = setOf(Frog::class,
                    Sample::class))
            val realm = Realm.open(config)
            Log.v("Successfully opened realm:" +
                    "${realm.configuration.name}")
            // :code-block-end:
            realm.close()
        }
    }

    @Test
    fun openARealmBuilderTest() {
        val REALM_NAME = getRandom()
        val PATH = TMP_PATH
        val KEY = ByteArray(64)

        runBlocking {
            // :code-block-start: open-a-realm-advanced
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .deleteRealmIfMigrationNeeded()
                .directory(PATH)
                .encryptionKey(KEY)
                .build()
            val realm = Realm.open(config)
            Log.v("Successfully opened realm:" +
                    realm.configuration.name
            )
            // :code-block-end:
            realm.close()
        }
    }

    @Test
    fun asyncWriteTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .directory(TMP_PATH)
                .build()
            val realm = Realm.open(config)
            Log.v("Successfully opened realm:" +
                    realm.configuration.name
            )
            // :code-block-start: write-async
            realm.write {
                // this: MutableRealm
                val sample = Sample()
                sample.stringField = "Sven"
                this.copyToRealm(sample)
            }
            // :code-block-end:
            realm.close()
        }
    }

    @Test
    fun syncWriteTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .directory(TMP_PATH)
                .build()
            val realm = Realm.open(config)
            Log.v("Successfully opened realm: ${realm.configuration.name}")
            // :code-block-start: write-sync
            realm.writeBlocking {
                // this: MutableRealm
                val sample = Sample()
                sample.stringField = "Sven"
                this.copyToRealm(sample)
            }
            // :code-block-end:
            realm.close()
        }
    }

    @Test
    fun queryTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .directory(TMP_PATH)
                .build()
            val realm = Realm.open(config)
            Log.v("Successfully opened realm: ${realm.configuration.name}")
            // :code-block-start: query-filters
            val samples: RealmResults<Sample> =
                realm.query<Sample>().find()

            val samplesThatBeginWithN:
                    RealmResults<Sample> =
                realm.query<Sample>(
                    "stringField BEGINSWITH 'N'"
                ).find()
            // :code-block-end:
            realm.close()
        }
    }

    @Test
    fun querySortLimitDistinctTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .directory(TMP_PATH)
                .build()
            val realm = Realm.open(config)
            Log.v("Successfully opened realm: ${realm.configuration.name}")
            // :code-block-start: query-sort-limit-distinct
            val aggregates: RealmResults<Sample> =
                realm.query<Sample>()
                    .distinct(Sample::stringField.name)
                    .sort(Sample::stringField.name,
                        Sort.ASCENDING)
                    .limit(2)
                    .find()
            // :code-block-end:
            realm.close()
        }
    }

    @Test
    fun deleteTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .directory(TMP_PATH)
                .build()
            val realm = Realm.open(config)
            Log.v("Successfully opened realm: ${realm.configuration.name}")
            realm.writeBlocking { // this: MutableRealm
                val sample = Sample()
                sample.stringField = "Sven"
                this.copyToRealm(sample)
                sample.stringField = "not sven"
                this.copyToRealm(sample)
            }
            // :code-block-start: deletes
            val sample: Sample? =
                realm.query<Sample>()
                    .first().find()

            // delete one object synchronously
            realm.writeBlocking {
                if (sample != null) {
                    findLatest(sample)
                        ?.also { delete(it) }
                }
            }

            // delete a query result asynchronously
            GlobalScope.launch {
                realm.write {
                    query<Sample>()
                        .first()
                        .find()
                        ?.also { delete(it) }
                }
            }
            // :code-block-end:
            realm.close()
        }
    }

    @Test
    fun notificationTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .directory(TMP_PATH)
                .build()
            val realm = Realm.open(config)
            Log.v("Successfully opened realm: ${realm.configuration.name}")
            realm.writeBlocking { // this: MutableRealm
                val sample = Sample()
                sample.stringField = "Sven"
                this.copyToRealm(sample)
                sample.stringField = "not sven"
                this.copyToRealm(sample)
            }

            val asyncQuery = async {
                // :code-block-start: notifications
                // in a coroutine or a suspend function
                realm.query<Sample>().asFlow().collect {
                        results: ResultsChange<Sample> ->
                    when (results) {
                        is InitialResults<Sample> -> {
                            // do nothing with the
                            // initial set of results
                        }
                        is UpdatedResults<Sample> -> {
                            // log change description
                            Log.v("Results changed. " +
                                "change ranges: " +
                                results.changeRanges +
                                ", insertion ranges: " +
                                results.insertionRanges +
                                ", deletion ranges: " +
                                results.deletionRanges
                            )
                        }
                    }
                }
                // :code-block-end:
            }
            asyncQuery.cancel()
            realm.close()
        }
    }

    @Test
    fun threadingTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .directory(TMP_PATH)
                .build()

            // :code-block-start: threading
            val realm = Realm.open(config)
            // :hide-start:
            realm.writeBlocking { // this: MutableRealm
                val sample = Sample()
                sample.stringField = "Sven"
                this.copyToRealm(sample)
                sample.stringField = "not sven"
                this.copyToRealm(sample)
            }
            // :hide-end:
            val sample: Sample? =
                realm.query<Sample>()
                    .first()
                    .find()

            launch(Dispatchers.Unconfined) {
                // can access the realm opened on
                // a different thread
                realm.query<Sample>().find()
                // can access realm object queried
                // on a different thread
                Log.v(sample!!.stringField)
            }.join()
            // :code-block-end:
            realm.close()
        }
    }

    @Test
    fun migrationTest() {
        val REALM_NAME = getRandom()

        runBlocking {
            val config = RealmConfiguration.Builder(
                setOf(Frog::class, Sample::class))
                .name(REALM_NAME)
                .directory(TMP_PATH)
                .build()

            // :code-block-start: migrations
            // A Realm migration that performs
            // automatic schema migration
            // and allows additional custom
            // migration of data.
            RealmConfiguration.Builder(
                schema = setOf(Sample::class))
                .migration(AutomaticSchemaMigration {
                        context:
                            AutomaticSchemaMigration.MigrationContext ->
                    val oldRealm:
                            DynamicRealm =
                        context.oldRealm
                    val newRealm:
                            DynamicMutableRealm =
                        context.newRealm
                    // dynamic realm gives access
                    // to realm data
                    // through a generic string
                    // based API
                    context.enumerate("Sample") {
                            oldObject:
                                DynamicRealmObject,
                            newObject:
                                DynamicMutableRealmObject? ->
                        newObject?.set("longField",
                                        42L)
                    }
                })
                .build()
            val realm = Realm.open(config)
            // :code-block-end:
            realm.close()
        }
    }
}